Updated pull request for trustedqsl

k1mu opened the pull request:
### Please confirm your submission meets all the criteria

- [X] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [X] My pull request follows the instructions at [App Submission][submission].
- [X] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [X] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [X] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. **Link:**
- [X] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
- [X] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*

[reqs]: https://github.com/flathub/flathub/wiki/App-Requirements
[maint]: https://github.com/flathub/flathub/wiki/App-Maintenance
[submission]: https://github.com/flathub/flathub/wiki/App-Submission

k1mu commented:
Corrective actions: removed unnecessary cleanup actions, added required configuration (icons, screenshot). Changed wx build to cmake-ninja.

hfiguiere commented on org.arrl.trustedqsl.json:
```diff
@@ -0,0 +1,148 @@
+    "command" : "tqsl",
+    "finish-args" : [
+	"--share=network",
+	"--socket=x11",
```
`share=ipc` is  needed for X11.

k1mu replied:
Added - I missed this in the last round.

hfiguiere commented on org.arrl.trustedqsl.json:
```diff
@@ -0,0 +1,148 @@
+     ],
+    "modules" : [
+        {
+            "name" : "openssl",
```
I will repeat what I said on the PR you closed: you do NOT need to build either openssl or curl.

k1mu replied:
Well, duh. You're absolutely correct - a build without those works just fine. 
I'm removing this.

hfiguiere commented on org.arrl.trustedqsl.json:
```diff
@@ -0,0 +1,148 @@
+	    ]
+        },
+        {
+            "name" : "curl",
```
repeating once again: curl is in the runtime.

k1mu replied:
Removed.

hfiguiere commented on org.arrl.trustedqsl.json:
```diff
@@ -0,0 +1,148 @@
+		    "-DCMAKE_BUILD_TYPE=Release",
+		    "-DwxUSE_UNICODE=Yes"
+	    ],
+	    "sources" : [
```
sources is usually last in the module by good practice. This applies everywhere.

k1mu replied:
Corrected - moved the sources blocks to the end.

hfiguiere commented on org.arrl.trustedqsl.json:
```diff
@@ -0,0 +1,148 @@
+	},
+	{
+	    "name" : "tqsl",
+	    "buildsystem" : "cmake",
```
it doesn't build with cmake-ninja ?

k1mu replied:
Yes, it does. Changed.

hfiguiere commented on org.arrl.trustedqsl.json:
```diff
@@ -0,0 +1,148 @@
+    "finish-args" : [
+	"--share=network",
+	"--socket=x11",
+	"--filesystem=home"
```
this is a broad filesystem permission.

k1mu replied:
TrustedQSL stores configuration data in $HOME/.tqsl and can be used to sign log files from other logging applications.  I don't think full filesystem access is needed, but the way this is used (primarily as a utility for loggers to sign and upload), home access is needed.

hfiguiere replied:
then you can use `--persist` instead.

ideally it would use xdg standard locations.

k1mu replied:
I'll try with --persist.

AsciiWolf replied:
`--persist=.tqsl` should work fine.

k1mu commented:
I've pushed an update based on the comments so far.

hfiguiere commented:
bot, build org.arrl.trustedqsl

hfiguiere commented:
tha app-id in the appstream file is incorrect.

k1mu commented:
I see the error, will correct and resubmit.

k1mu commented:
Trying with --persist versus --filesystem=home effectively makes this unusable. None of the configuration data (certificates, locations, etc.) is visible from the home folder and there's no way to import any of the necessary data. Opening files from outside of the container also fails, which is the major purpose of the app - to take a file exported by another application, read that for contact data, sign it, then save or upload the resulting container. 

I understand the desire to restrict what data applications have access to, but this can't run with container-only access.

k1mu commented:
I've tried a clean build to verify the behavior with --persist=.tqsl and that's not going to work.
Let me explain this application and what it does, with hopes of making the case for --filesystem=home.

This is an application that has two facets - a GUI tool and a command line tool.
The GUI manages a PKI for the user - generating a certificate request and either saving it for later upload, or directly uploading that request.  The central system performs an identity verification step (which can include using the mail to validate that the requester is indeed the license holder) and issues a signed PKCS6 file back to the user once they're verified. There's some unique certificate extensions to indicate what "country" someone is  from (the country that issued their ham license) as well as a separate set of validity dates. (The certificate might expire in 3 years, but that ham license has a separate validity period that could be just a small number of days.) This is why there's a specialized PKI.
In addition, the GUI allows managing the data about the physical location of the operator - State, County, etc.

That gui management is relatively infrequent, but the data needs to be persistent and portable between computer systems. There's a backup/restore mechanism to allow populating a new computer with the operator's details.

Then, given that setup, a user submits a file of radio contacts in a standardized form (ADIF - Amateur Data Interchange Format).  those contacts are augmented with a digital signature verifying that the operator made that radio contact. The resulting signed file is either saved for later upload, or directly uploaded if there's an Internet connection.

The centralized server, Logbook of the World, matches contacts - station "A" says they contacted station "B" at a given date, time, and other details. Matches are generated and can be used for awards - operating a certain number of countries.

The majority of the time, radio contacts are managed by completely separate logging programs. Those emit ADIF-format logs and use the application's command line interface to digitally sign and optionally upload contacts. It's fairly rare for the users to manually upload themselves (but there's the ability to do that, as well as a GUI menu for creating and editing ADIF files.)

All of that means that an arbitrary program has to be able to pass "tqsl random-file.adif" and have it read, process, and upload the result. So TQSL has to be able to open files wherever that application put them.  I've tried with the mentioned build with "-persist=.tqsl" and I'm not able to open any files outside of that directory.  In addition, the existing ".tqsl" folder isn't being mapped, so  the existing configuration is lost. I can't even copy a backup package to that folder as it's not visible outside of the container.

This application is ancient - initially started in 2001.

tl;dr - think of this as a filter that takes input files and transforms them. I've got to be able to get input from almost anywhere; While I can state that there's a limitation for the flatpak build that any input files have to be somewhere under $HOME I can't expect several logging programs to adapt to a highly restrictive folder structure.

That all being said, I've been wildly off base already here. I am not trying to be argumentative and would be happy to discuss or be proven wrong yet again.

hfiguiere commented:
> Let me explain this application and what it does, with hopes of making the case for --filesystem=home.

Then `--filesystem=~/.tqsl`

k1mu commented:
> > Let me explain this application and what it does, with hopes of making the case for --filesystem=home.
> 
> Then `--filesystem=~/.tqsl`

That does get me to the existing configuration data, so that's a start.  However, wxWidgets stores config data in a dotfile off of home, so "--filesystem=~/.tqslapp" is needed as well. But that starts a slippery slope. 

Then how is the application supposed to open file ~/cqrlog/k1mu.adif when that application saves the file?
If the user saves a certificate as ~/callsign.tq6, how can they import it?

I can't dictate this big of a change on the dependent programs. And  having this be so restrictive on Linux versus MacOS/BSD/Windows would be a big downgrade - as well as being a significant regression from the packages that some distributions already have in place (many of which aren't being maintained, thus the idea of a Flatpak build so I can manage this.)
-

k1mu commented:
Another option here that might be worth trying is to add the xdg-documents and xdg-downloads folders. I haven't tried that to see how it works, but that may be worth considering.

hfiguiere commented:
AppStream fail to validate.
```
builddir/files/share/appdata/org.arrl.trustedqsl.appdata.xml: FAILED:
• tag-missing           : <release> required
Validation of files failed
```

hfiguiere commented:
bot, build org.arrl.trustedqsl

hfiguiere commented:
the icon is not found.

hfiguiere commented:
```
-- Installing: /app/share/app-info/hicolor/icons/128x128/org.arrl.trustedqsl.png
-- Installing: /app/share/app-info/hicolor/icons/64x64/org.arrl.trustedqsl.png
-- Installing: /app/share/app-info/hicolor/icons/48x48/org.arrl.trustedqsl.png
-- Installing: /app/share/app-info/hicolor/icons/32x32/org.arrl.trustedqsl.png
-- Installing: /app/share/app-info/hicolor/icons/16x16/org.arrl.trustedqsl.png
```

This is  very wrong as icon path.

barthalion commented:
bot, build org.arrl.trustedqsl

barthalion commented:
Failed `Check for 128x128 icon` typically means a mismatch somewhere between the appdata, desktop and icon files. The compose command first reads the appdata file, uses ID from it (alternatively, refers to `<launchable>`) to look up the desktop file, from which it reads the `Icon` field. Everything has to match the app ID used in the manifest.

hfiguiere commented:
I don't see a .desktop file installed.

k1mu commented:
Thanks - that's easily fixed.
I've run the desktop validator against the desktop file, removed the icon spec from the metadata XML file, and added to the desktop. The icon is now being installed to the "right" directory.  Will update the package once I check the build output.

hfiguiere commented:
bot, build org.arrl.trustedqsl

barthalion commented:
It segfaults on my end:

```
[bpiotrowski@purei ~]$ coredumpctl info
           PID: 37484 (tqsl)
           UID: 1000 (bpiotrowski)
           GID: 1000 (bpiotrowski)
        Signal: 11 (SEGV)
     Timestamp: Fri 2023-05-12 16:14:51 CEST (30s ago)
  Command Line: tqsl
    Executable: /app/bin/tqsl
 Control Group: /user.slice/user-1000.slice/user@1000.service/app.slice/app-flatpak-org.arrl.trustedqsl-37467.scope
          Unit: user@1000.service
     User Unit: app-flatpak-org.arrl.trustedqsl-37467.scope
         Slice: user-1000.slice
     Owner UID: 1000 (bpiotrowski)
       Boot ID: c75385931ced49a58ed473a5a22c2a09
    Machine ID: b245857f888a412b840acce4c4cc3577
      Hostname: purei
       Storage: /var/lib/systemd/coredump/core.tqsl.1000.c75385931ced49a58ed473a5a22c2a09.37484.1683900891000000.zst (present)
  Size on Disk: 1.4M
       Message: Process 37484 (tqsl) of user 1000 dumped core.
                
                Stack trace of thread 2:
                #0  0x00007f9a57e7bc35 n/a (/usr/lib/x86_64-linux-gnu/libc.so.6 + 0x7bc35)
                #1  0x0000555958df95c0 n/a (/app/bin/tqsl + 0x14f5c0)
                #2  0x00007f9a58790538 n/a (/app/lib/libwx_baseu-3.2.so.0.2.1 + 0x190538)
                #3  0x0000555958d8ad34 n/a (/app/bin/tqsl + 0xe0d34)
                #4  0x00007f9a57e2954a n/a (/usr/lib/x86_64-linux-gnu/libc.so.6 + 0x2954a)
                #5  0x00007f9a57e2960b n/a (/usr/lib/x86_64-linux-gnu/libc.so.6 + 0x2960b)
                #6  0x0000555958d8fb75 n/a (/app/bin/tqsl + 0xe5b75)
                ELF object binary architecture: AMD x86-64
```

k1mu commented:
The segfault is because there was insufficient checking of the code that creates the $HOME/.tqsl folder - when unable to do that, it'd eventually cause a null pointer dereference. A fix has been committed to the repo, will be pushing an update shortly.

barthalion commented:
bot, build org.arrl.trustedqsl

barthalion commented:
/merge
